### PR TITLE
feat: added custom segments app

### PR DESCRIPTION
**Pull request by amychisholm03**
This addition aims to move the examples in node-newrelic/examples/api/segments into a single stand-alone application. The application uses the fastify web framework to showcase how segments can be used in a variety of use cases: callback-based, promise-based, asynchronously, and synchronously.

### CODE REVIEW

File: custom-instrumentation/segments/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+   cp env.sample .env
+   # Fill out `NEW_RELIC_LICENSE_KEY` in .env and save 
+   # Start the application
+   npm start
+   ```
+4. The application will automatically start running background transactions with custom segments inside them. Modify index.js to include/omit more transactions/segments.
+
+## Exploring Telemetry
+
+After sending a few requests, navigate to your application in APM & Services. Locate the 'Example Custom Segments App' service and then select 'Distirbuted tracing' on the left side of your screen. Then, select one of the trace groups that will contain your custom segment and selct a single trace. If you expand the transaction (click 'Show in-process spans'), you should see your custom segment. Click on it to see its details.
```
Comment (jsumners-nr) [inline custom-instrumentation/segments/README.md:24]:
```suggestion
After sending a few requests, navigate to your application in APM & Services. Locate the 'Example Custom Segments App' service and then select 'Distirbuted tracing' on the left side of your screen. Then, select one of the trace groups that will contain your custom segment and select a single trace. If you expand the transaction (click 'Show in-process spans'), you should see your custom segment. Click on it to see its details.
```

File: custom-instrumentation/segments/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+![1721223889308](image/README/1721223889308.png)
+
+To query your segment data, you can run a query like the following:›
+
+```NRQL
+SELECT count(*) FROM Span FACET name WHERE name = 'myCustomSegment'
+```
+
+![1721224323000](image/README/1721224323000.png)
```
Comment (jsumners-nr) [inline custom-instrumentation/segments/README.md:34]:
```suggestion
![1721223889308](./image/README/1721223889308.png)

To query your segment data, you can run a query like the following:›

```NRQL
SELECT count(*) FROM Span FACET name WHERE name = 'myCustomSegment'
```


```

File: custom-instrumentation/segments/index.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            // The handler is the function that will be wrapped with the new segment. When
+            // a callback is provided, the segment timing will end when the callback is
+            // called.
+
+            newrelic.startSegment('myCallbackSegment', false, someTask, function cb(err, output) {
+                // Handle the error and output as appropriate.
+                tx.end()
+                resolve(output)
+            })
+        })
```
Comment (jsumners-nr) [inline custom-instrumentation/segments/index.js:63]: I think we should be consistent with the comment styles. My personal preference is to reserve multiline comments for doc blocks, e.g.:

```js
/**
 * Documentation for the function.
 *
 * @param {string} foo
 * 
 * @returns {string}
 */
function foo (foo) {
	return foo + '-foo'
}
```

For explanations in code, I like to use single line comments:

```js
/**
 * Docs
 */
function foo () {
  // This function's implementation is not very obvious.
	// We had to foo the baz to get bar.
}
```

See https://github.com/jsumners/unit-test-patterns-and-strategies/blob/master/lessons/001-basic-testing/index.test.js (and other files in that repo) for fleshed out examples.

Comment (amychisholm03) [reply custom-instrumentation/segments/index.js:63]: Sounds good! I just copied these comments over from the old examples.

**Review by bizob2828 — CHANGES_REQUESTED** [review]
The intent of replacing these examples from node-newrelic is to wrap them in a webframework so you don't have to run in background transaction. The only example that should be running in background transactions is the example to show how to use that api method.  This example is showing how to record custom segments

File: custom-instrumentation/segments/index.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    })
+}
+
+//example2-promises
+function promiseSegment() {
```
Comment (bizob2828) [inline custom-instrumentation/segments/index.js:68]: you may want to put the example functions that are getting run in a different file for better organization as well

File: custom-instrumentation/segments/index.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            // `startSegment()` takes a segment name, a boolean if a metric should be
+            // created for this segment, the handler function, and an optional callback.
+            // The handler is the function that will be wrapped with the new segment. If
+            // a promise is returned from the handler, the segment's ending will be tied
+            // to that promise resolving or rejecting.
+
+            return newrelic.startSegment('myPromiseSegment', false, someTask).then(function thenAfter(output) {
+                tx.end()
+                resolve(output)
+            })
```
Comment (bizob2828) [inline custom-instrumentation/segments/index.js:112]:
```suggestion
            const output = await newrelic.startSegment('myPromiseSegment', false, someTask)
             tx.end()
             resolve(output)
            })
```

Comment (bizob2828) [reply custom-instrumentation/segments/index.js:112]: oh n/m, this is intentional

File: custom-instrumentation/segments/index.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+const newrelic = require('newrelic')
+
+//example1-callbacks
+function callbackSegment() {
+    return new Promise((resolve) => {
```
Comment (bizob2828) [inline custom-instrumentation/segments/index.js:11]: this doesn't have to be here if you update this to use a webframework.

File: custom-instrumentation/segments/functions.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const { someCbTask } = require('./functions/callback.js')
```
Comment (bizob2828) [inline custom-instrumentation/segments/functions.js:6]: this is fine but you can put this in an index.js within `functions` folder and you can still require it as `require('./functions')`

File: custom-instrumentation/segments/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+   ```
+   npm install
+   cp env.sample .env
+   # Fill out `NEW_RELIC_LICENSE_KEY` in .env and save 
+   # Start the application
+   npm start
+   ```
+4. Make requests to the application. Route names include: 'callback-segment', 'promise-segment', 'async-segment', and 'sync-assign-segment'.
+   ```
+   curl -X POST http://localhost:3000/custom-segment
```
Comment (bizob2828) [inline custom-instrumentation/segments/README.md:22]: this URI doesn't exist. 
```suggestion
   curl -X POST http://localhost:3000/callback-segment
   curl -X POST http://localhost:3000/promise-segment
   curl -X POST http://localhost:3000/async-segment
   curl -X POST http://localhost:3000/sync-assign-segment
```

File: custom-instrumentation/segments/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+## Exploring Telemetry
+
+After sending a few requests, navigate to your application in APM & Services. Locate the 'Example Custom Segments App' service and then select 'Distributed tracing' on the left side of your screen. Then, select one of the trace groups that will contain your custom segment and select a single trace. If you expand the transaction (click 'Show in-process spans'), you should see your custom segment. Click on it to see its details.
+
+![1721223889308](./image/README/1721223889308.png)
+
+To query your segment data, you can run a query like the following:›
+
+```NRQL
+SELECT count(*) FROM Span FACET name WHERE name = 'myCustomSegment'
```
Comment (bizob2828) [inline custom-instrumentation/segments/README.md:34]:
```suggestion
SELECT count(*) FROM Span FACET name WHERE name = 'myCallbackSegment'
```

File: custom-instrumentation/segments/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+   curl -X POST http://localhost:3000/custom-segment
+   ```
+
+## Exploring Telemetry
+
+After sending a few requests, navigate to your application in APM & Services. Locate the 'Example Custom Segments App' service and then select 'Distributed tracing' on the left side of your screen. Then, select one of the trace groups that will contain your custom segment and select a single trace. If you expand the transaction (click 'Show in-process spans'), you should see your custom segment. Click on it to see its details.
+
+![1721223889308](./image/README/1721223889308.png)
+
+To query your segment data, you can run a query like the following:›
```
Comment (bizob2828) [inline custom-instrumentation/segments/README.md:31]:
```suggestion
To query your segment data, you can run a query for the custom segments(myCallbackSegment, myPromiseSegment, myAsyncSegment, mySyncAssignSegment) like the following:
```

**Review by bizob2828 — CHANGES_REQUESTED** [review]
verified the app works as expected. just a few suggestions in the README and an alternate way to require the functions

**Review by bizob2828 — APPROVED** [review]
Nice work

**Pull request merged**